internal/testutil/mig/test_case_test.go

```diff
@@ -0,0 +1,114 @@
+package mig_test
+
+import (
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-testing/helper/resource"
+	"github.com/hashicorp/terraform-plugin-testing/terraform"
+	"github.com/mongodb/terraform-provider-mongodbatlas/internal/testutil/acc"
+	"github.com/mongodb/terraform-provider-mongodbatlas/internal/testutil/mig"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestConvertToMigration(t *testing.T) {
+	var (
+		preCheckCalled     = false
+		checkDestroyCalled = false
+		config             = "someTerraformConfig"
+	)
+	preCheck := func() {
+		preCheckCalled = true
+	}
+	firstStep := resource.TestStep{
+		Config: config,
+		Check:  resource.TestCheckResourceAttrSet("someTarget", "someAttribute"),
+	}
+
+	asserter := assert.New(t)
+
+	convertAndCall := func(test resource.TestCase) resource.TestCase {
+		newTest := mig.ConvertToMigrationTest(t, &test)
+		newTest.PreCheck()
+		if newTest.CheckDestroy != nil {
+			asserter.NoError(newTest.CheckDestroy(nil))
+		}
+		return newTest
+	}
+	defaultAssertions := func(test resource.TestCase) {
+		t.Helper()
+		asserter.Len(test.Steps, 2, "Expected 2 steps (one extra test step)")
+		newFirstStep := test.Steps[0]
+		asserter.Equal(config, newFirstStep.Config)
+
+		planStep := test.Steps[1]
+		asserter.Equal(mig.TestStepCheckEmptyPlan(config), planStep)
+	}
+
+	t.Run("normal call with check and destroy", func(t *testing.T) {
+		checkDestroy := func(*terraform.State) error {
+			checkDestroyCalled = true
+			return nil
+		}
+		test := convertAndCall(resource.TestCase{
+			PreCheck:     preCheck,
+			CheckDestroy: checkDestroy,
+			Steps: []resource.TestStep{
+				firstStep,
+			},
+		})
+		asserter.True(preCheckCalled)
+		asserter.True(checkDestroyCalled)
+		defaultAssertions(test)
+	})
+
+	t.Run("checkDestroy=nil has no panic", func(t *testing.T) {
+		test := convertAndCall(resource.TestCase{
+			PreCheck: preCheck,
+			Steps: []resource.TestStep{
+				firstStep,
+			},
+		})
+		defaultAssertions(test)
+	})
+
+	t.Run("more than 1 step uses only 1 step", func(t *testing.T) {
+		test := convertAndCall(resource.TestCase{
+			PreCheck: preCheck,
+			Steps: []resource.TestStep{
+				firstStep,
+				{
+					Config: "differentConfig",
+					Check:  resource.TestCheckResourceAttrSet("target", "attribute"),
+				},
+			},
+		})
+		defaultAssertions(test)
+	})
+	// ConvertToMigrationTestUseExternalProvider
```

Comment:
Put both unit test for: `ConvertToMigrationTest` and `ConvertToMigrationTestUseExternalProvider` in the same test as they are reusing some internal functions.
However, would love to see suggestions of how to make this test cleaner

Reply:
it's quite a big test, don't know if some nested tests could be extracted to functions, or if a table-based test would help here